File: algorithms/2_medium/F/00373_find_k_pairs_with_smallest_sums/heap.cpp
class Solution {
    struct Cmp {
        bool operator()(const tuple<int, int, int>& a, const tuple<int, int, int>& b) {
            return std::get<0>(a) > std::get<0>(b);
        }
    };
public:
    vector<vector<int>> kSmallestPairs(vector<int>& nums1, vector<int>& nums2, int k) {
        assert(
            k >= 1 && k <= std::ssize(nums1) * std::ssize(nums2) &&
            nums1.size() > 0 && nums1.size() <= 100'000 &&
            nums2.size() > 0 && nums2.size() <= 100'000
        );

        const int len1 = static_cast<int>(std::ssize(nums1));
        const int len2 = static_cast<int>(std::ssize(nums2));
        auto heap = priority_queue<tuple<int, int, int>, vector<tuple<int, int, int>>, Cmp>();
        const int heap_len = std::min(k, len1);
        for (int i = 0; i < heap_len; i += 1) {
            heap.push({ nums1[i] + nums2[0], i, 0 });
        }
        auto res = vector<vector<int>>();
        res.reserve(k);

        for (int step = 0; step < k; step += 1) {
            const auto [sum, idx1, idx2] = heap.top();
            heap.pop();
            res.push_back(vector<int>{ nums1[idx1], nums2[idx2] });
            if (idx2 + 1 < len2) {
                heap.push({ nums1[idx1] + nums2[idx2 + 1], idx1, idx2 + 1 });
            }
        }

        return res;
    }
};

File: algorithms/2_medium/F/00373_find_k_pairs_with_smallest_sums/heap.rs
use std::collections::BinaryHeap;
use std::cmp::Reverse;

impl Solution {
    pub fn k_smallest_pairs(nums1: Vec<i32>, nums2: Vec<i32>, k: i32) -> Vec<Vec<i32>> {
        assert!(
            k >= 1 && (k as usize) <= nums1.len() * nums2.len() &&
            nums1.len() > 0 && nums1.len() <= 100_000 &&
            nums2.len() > 0 && nums2.len() <= 100_000
        );

        let len1 = nums1.len();
        let k = k as usize;
        let mut heap = BinaryHeap::with_capacity(k.min(len1));
        for i in 0..(k.min(len1)) {
            heap.push(Reverse((nums1[i] + nums2[0], i as i32, 0_i32)));
        }
        let mut res: Vec<Vec<i32>> = Vec::with_capacity(k);
        let len2 = nums2.len() as i32;

        for _ in 0..k {
            let (Reverse((sum, idx1, idx2))) = heap.pop().unwrap();
            res.push(vec![nums1[idx1 as usize], nums2[idx2 as usize]]);
            if idx2 + 1 < len2 {
                heap.push(Reverse((nums1[idx1 as usize] + nums2[(idx2 + 1) as usize], idx1, idx2 + 1)));
            }
        }

        res
    }
}
